dotnet/src/Planners/Planners.Core/Handlebars/Extensions/SKParameterMetadataExtensions.cs

```diff
@@ -0,0 +1,146 @@
+
+#pragma warning disable IDE0130 // Namespace does not match folder structure
+namespace Microsoft.SemanticKernel.Planning.Handlebars.Extensions;
+#pragma warning restore IDE0130
+
+internal static class SKParameterMetadataExtensions
+{
+    /// <summary>
+    /// Checks if type is primitive or string
+    /// </summary>
+    public static bool isPrimitiveOrStringType(Type type) => type.IsPrimitive || type == typeof(string);
+
+    /// <summary>
+    /// Converts non-primitive types to a data class definition and returns a hash set of complex type metadata.
+    /// Complex types will become a data class.
+    /// If there are nested complex types, the nested complex type will also be returned.
+    /// Example:
+    /// Complex type:
+    /// class ComplexType:
+    ///    propertyA: int
+    ///    propertyB: str
+    ///    propertyC: PropertyC
+    /// </summary>
+    public static HashSet<HandlebarsParameterTypeMetadata> ToHandlebarsParameterTypeMetadata(this Type type)
+    {
+        var parameterTypes = new HashSet<HandlebarsParameterTypeMetadata>();
+        if (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(Task<>))
+        {
+            // Async return type - need to extract the actual return type
+            var actualReturnType = type.GenericTypeArguments[0]; // Actual Return Type
+            var returnTypeProperties = actualReturnType.GetProperties();
+
+            if (!isPrimitiveOrStringType(actualReturnType) && returnTypeProperties.Length is not 0)
+            {
+                parameterTypes.Add(new HandlebarsParameterTypeMetadata()
+                {
+                    Name = actualReturnType.Name,
+                    IsComplex = true,
+                    Properties = returnTypeProperties.Select(p => new SKParameterMetadata(p.Name) { ParameterType = p.PropertyType }).ToList()
+                });
+
+                parameterTypes.AddNestedComplexTypes(returnTypeProperties);
+            }
+        }
+        else if (type.IsClass && type != typeof(string))
+        {
+            // Class
+            var properties = type.GetProperties();
+
+            parameterTypes.Add(new HandlebarsParameterTypeMetadata()
+            {
+                Name = type.Name,
+                IsComplex = properties.Length is not 0,
+                Properties = properties.Select(p => new SKParameterMetadata(p.Name) { ParameterType = p.PropertyType }).ToList()
+            });
+
+            parameterTypes.AddNestedComplexTypes(properties);
+        }
+
+        return parameterTypes;
+    }
+
+    private static void AddNestedComplexTypes(this HashSet<HandlebarsParameterTypeMetadata> parameterTypes, PropertyInfo[] properties)
+    {
+        // Add nested complex types
+        foreach (var property in properties)
+        {
+            parameterTypes.UnionWith(property.PropertyType.ToHandlebarsParameterTypeMetadata());
+        }
+    }
+
+    public static bool isPrimitiveOrStringType(string type) =>
```

Comment:
Do we need this overload of the `isPrimitiveOrStringType` method taking into account we already declared one above? 
Can we keep the two methods next to each other?

Reply:
Yes one takes a `Type` parameter and the other takes a stringified type (type serialized in JSON Schema). Moved them to be next to each other

![image](https://github.com/microsoft/semantic-kernel/assets/125500434/8fd57245-c1c9-4bc0-99d4-ccd2a05525fc)